src/gnu/prolog/term/Rational.java

```diff
@@ -0,0 +1,248 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA  02111-1307, USA. The text of license can be also found
+ * at http://www.gnu.org/copyleft/lgpl.html
+ */
+
+package gnu.prolog.term;
+import java.math.BigInteger;
+import java.math.BigDecimal;
+import java.math.MathContext;
+
+public class Rational implements Comparable
+{
+	private MathContext mc = MathContext.DECIMAL64;
+	private BigInteger numerator;
+	private BigInteger denominator;
+	public static final BigInteger INTEGER_MIN = BigInteger.valueOf(Integer.MIN_VALUE);
+	public static final BigInteger INTEGER_MAX = BigInteger.valueOf(Integer.MAX_VALUE);
+
+	public Rational(BigInteger numerator, BigInteger denominator)
+	{
+		if (denominator.equals(BigInteger.ZERO))
+		{
+			throw new ArithmeticException("Division by zero");
+		}
+		BigInteger gcd = numerator.gcd(denominator);
+		if (denominator.signum() == -1)
+		{
+			// multiply by -1/-1
+			numerator = numerator.negate();
+			denominator = denominator.negate();
+		}
+
+		if (gcd.equals(BigInteger.ONE))
+		{
+			// Save 2 division operations here
+			this.numerator = numerator;
+			this.denominator = denominator;
+		}
+		else
+		{
+			this.numerator = numerator.divide(gcd);
+			this.denominator = denominator.divide(gcd);
+		}
+	}
+
+	@Override
+	public boolean equals(Object o)
+	{
+		return (o instanceof Rational &&
+			((Rational)o).numerator.equals(numerator) &&
+			((Rational)o).denominator.equals(denominator));
+	}
+
+	@Override
+	public int compareTo(Object o)
+	{
+		if (o instanceof Rational)
+		{
+			Rational subtrahend = (Rational)o;
+			return numerator.multiply(subtrahend.denominator).subtract(subtrahend.numerator.multiply(denominator)).signum();
+		}
+		return hashCode() - o.hashCode();
+	}
+
+	@Override
+	public String toString()
+	{
+		// Note that this just prints out the value to 20dp
+		// We do have more precision than that, but we have to draw the line somewhere when asked to print
+		return numerator.toString() + " rdiv " + denominator.toString();
+		//return new BigDecimal(numerator).divide(new BigDecimal(denominator), mc).toString();
+	}
+
+	public double doubleValue()
+	{
+		return new BigDecimal(numerator).divide(new BigDecimal(denominator), mc).doubleValue();
+	}
+
+	public static Rational get(int val)
+	{
+		return new Rational(BigInteger.valueOf(val), BigInteger.valueOf(1));
+	}
+
+	public static Rational get(BigInteger val)
+	{
+		return new Rational(val, BigInteger.valueOf(1));
+	}
+
+        private static final double DBL_EPSILON = 0.00000000000000022204;
+        public static Rational getApproximate(double val)
+        {
+                // This is borrowed directly from SWI Prolog
```

Comment:
I'm not sure. The code is inspired very directly by SWI Prolog, but the same algorithm is also implemented in YAP, and apparently in Common LISP. I've found another copy of the paper, at https://www.cs.duke.edu/brd/papers/src-papers/rotations-scg92.pdf and the algorithm is presented in a very similar format in figure 3.